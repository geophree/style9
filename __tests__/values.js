/* eslint-env jest */
const compile = require('./compile.js');

it('compiles', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    color: 'blue'
  }
});
styles('default');
  `;
  const { code } = compile(input);
  expect(code).not.toBe(input);
});

it('converts to pixels', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    paddingLeft: 2
  }
});
styles('default');
  `;
  const { styles } = compile(input);

  expect(styles).toMatchSnapshot();
});

it('does not convert to pixels', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    opacity: 1,
    '--custom': 2
  }
});
styles('default');
  `;
  const { styles } = compile(input);

  expect(styles).toMatchSnapshot();
});

it('expands shorthand', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    padding: '1rem'
  }
});
styles('default');
  `;
  const { code, styles } = compile(input);

  expect(code).toMatchSnapshot();
  expect(styles).toMatchSnapshot();
});

it('does not override longhand', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    padding: '1rem',
    paddingLeft: '2rem'
  }
});
styles('default');
  `;
  const { code, styles } = compile(input);

  expect(code).toMatchSnapshot();
  expect(styles).toMatchSnapshot();
});

it('converts fontSize pixels', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    fontSize: 14
  }
});
styles('default');
  `;
  const { styles } = compile(input);

  expect(styles).toMatchSnapshot();
});

it('accepts an array', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    textDecorationLine: ['underline', 'overline']
  }
});
styles('default');
  `;
  const { styles } = compile(input);

  expect(styles).toMatchSnapshot();
});

it('supports constants', () => {
  const input = `
import style9 from 'style9';
const BLUE = 'blue';
const styles = style9.create({
  default: {
    color: BLUE
  }
});
styles('default');
  `;
  const { styles } = compile(input);

  expect(styles).toMatchSnapshot();
});

it('kebab-cases non-custom property names', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    borderTopColor: 'blue',
    '--customProp': 'red'
  }
});
styles('default');
  `;
  const { code, styles } = compile(input);

  expect(styles).toMatchSnapshot();
});

it('removes unused styles', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    color: 'blue'
  }
});
  `;
  const { code, styles } = compile(input);

  expect(code).toMatchSnapshot();
  expect(styles).toMatchSnapshot();
});

it('supports static bracket access', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    color: 'blue'
  },
  red: {
    color: 'red'
  }
});
styles['default']
  `;
  const { code, styles } = compile(input);

  expect(code).toMatchSnapshot();
  expect(styles).toMatchSnapshot();
});

it('supports dynamic bracket access', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  blue: {
    color: 'blue'
  },
  red: {
    color: 'red'
  }
});
styles[blue]
  `;
  const { code, styles } = compile(input);

  expect(code).toMatchSnapshot();
  expect(styles).toMatchSnapshot();
});

it('supports arrow function', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    color: 'blue'
  }
});
const get = state => styles(state && 'default');
  `;
  const { code, styles } = compile(input);

  expect(code).toMatchSnapshot();
  expect(styles).toMatchSnapshot();
});

it('does not work without declaration', () => {
  const input = `
import style9 from 'style9';
style9.create({
  default: {
    color: 'blue'
  }
});
  `;
  expect(() => compile(input)).toThrow();
});

it('only supports Member- and CallExpression on styles', () => {
  const input = `
import style9 from 'style9';
const styles = style9.create({
  default: {
    color: 'blue'
  }
});
foo(styles);
  `;
  expect(() => compile(input)).toThrow();
});

it('throws on non-existing property', () => {
  const input = `
import style9 from 'style9';
style9.foo;
  `;
  expect(() => compile(input)).toThrow();
});
