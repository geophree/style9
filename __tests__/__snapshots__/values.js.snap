// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts an array 1`] = `".cdd1pm5{text-decoration-line:underline overline}"`;

exports[`converts fontSize pixels 1`] = `".c1eftxfu{font-size:0.875rem}"`;

exports[`converts to pixels 1`] = `".csaw36d{padding-left:2px}"`;

exports[`does not converts to pixels 1`] = `".cu2kwdz{opacity:1}"`;

exports[`does not override longhand 1`] = `
"import style9 from 'style9';
const styles = {
  default: {
    paddingTop: \\"c1xiyxc9\\",
    paddingRight: \\"c1wgquni\\",
    paddingBottom: \\"c1s7zyp4\\",
    paddingLeft: \\"c17qt1w7\\"
  }
};
\\"c1xiyxc9 c1wgquni c1s7zyp4 c17qt1w7 \\";"
`;

exports[`does not override longhand 2`] = `".c1xiyxc9{padding-top:1rem}.c1wgquni{padding-right:1rem}.c1s7zyp4{padding-bottom:1rem}.c17qt1w7{padding-left:2rem}"`;

exports[`expands shorthand 1`] = `
"import style9 from 'style9';
const styles = {
  default: {
    paddingTop: \\"c1xiyxc9\\",
    paddingRight: \\"c1wgquni\\",
    paddingBottom: \\"c1s7zyp4\\",
    paddingLeft: \\"cc3nx7y\\"
  }
};
\\"c1xiyxc9 c1wgquni c1s7zyp4 cc3nx7y \\";"
`;

exports[`expands shorthand 2`] = `".c1xiyxc9{padding-top:1rem}.c1wgquni{padding-right:1rem}.c1s7zyp4{padding-bottom:1rem}.cc3nx7y{padding-left:1rem}"`;

exports[`removes unused styles 1`] = `
"import style9 from 'style9';
const styles = {};"
`;

exports[`removes unused styles 2`] = `""`;

exports[`supports arrow function 1`] = `
"import style9 from 'style9';
const styles = {
  default: {
    color: \\"c1r9f2e5\\"
  }
};

const get = state => state ? \\"c1r9f2e5 \\" : \\"\\";"
`;

exports[`supports arrow function 2`] = `".c1r9f2e5{color:blue}"`;

exports[`supports constants 1`] = `".c1r9f2e5{color:blue}"`;

exports[`supports dynamic bracket access 1`] = `
"import style9 from 'style9';
const styles = {
  blue: {
    color: \\"c1r9f2e5\\"
  },
  red: {
    color: \\"cyyg6ey\\"
  }
};
styles[blue];"
`;

exports[`supports dynamic bracket access 2`] = `".c1r9f2e5{color:blue}.cyyg6ey{color:red}"`;

exports[`supports static bracket access 1`] = `
"import style9 from 'style9';
const styles = {
  default: {
    color: \\"c1r9f2e5\\"
  }
};
styles['default'];"
`;

exports[`supports static bracket access 2`] = `".c1r9f2e5{color:blue}"`;
